File: subprograms/EdgeRDE.py
#!/usr/bin/env python3

"""
EdgeRDE — edgeR-based Differential Expression Analysis
SalmonStreamer Pipeline Module

Wraps src/R_src/edger_de.R, which implements a quasi-likelihood edgeR pipeline
with TMM normalisation, adaptive filterByExpr filtering, all pairwise contrasts,
PCA, sample-correlation heatmap, volcano plots, and a summary report.

Metadata file format (tab-separated, --metadata-file):
  REQUIRED COLUMNS:
    sample_name    Must match count matrix column headers exactly
                   (after any --sample-suffix stripping).
    group          Experimental group label used for contrasts
                   (e.g. "SpeciesA_Leaf", "SpeciesB_Bud").
                   ALTERNATIVE: provide 'species' + 'tissue' columns instead;
                   group will be derived as <species>_<tissue>.

  OPTIONAL COLUMNS (ignored by edger_de.R, may be present):
    species, tissue, platform, library, flowcell, lane, replicate, ...

CLI groups (--group-samples, used when --metadata-file is absent):
  Format:  GroupLabel:sample1,sample2,sample3  (space-separated per group)
  Example: --group-samples SpeciesA:s1,s2,s3 SpeciesB:s4,s5,s6
  The Python wrapper converts this into a temporary metadata file and passes
  it to the R script; --metadata-file takes priority if both are supplied.

@Author: Luis Javier Madrigal-Roca
@Date:   2026-04-30
"""

import argparse
import os
import sys
import subprocess
import tempfile
import csv
import re


def validate_expression_file(filepath):
    """
    Validate expression file format and contents.
    
    Parameters
    ----------
    filepath : str
        Path to expression file.
    
    Raises
    ------
    ValueError
        If file format is invalid or contains errors.
    """
    try:
        with open(filepath, 'r') as f:
            lines = f.readlines()
    except IOError as e:
        raise ValueError(f"Cannot read expression file: {e}")
    
    if len(lines) < 2:
        raise ValueError(
            "Expression file must have at least 2 rows (header + 1 gene). "
            f"Found {len(lines)} rows."
        )
    
    # Parse header
    try:
        header = lines[0].strip().split('\t')
    except Exception as e:
        raise ValueError(f"Cannot parse expression file header: {e}")
    
    if len(header) < 2:
        raise ValueError(
            f"Expression file must have at least 2 columns (gene_id + 1 sample). "
            f"Found {len(header)} columns."
        )
    
    # Check that all value columns are numeric
    for i, line in enumerate(lines[1:5], start=2):  # Check first few data rows
        try:
            fields = line.strip().split('\t')
            if len(fields) != len(header):
                raise ValueError(
                    f"Row {i} has {len(fields)} columns but header has {len(header)}"
                )
            # Try to convert all count columns to numbers (accept int or float)
            for count_str in fields[1:]:
                try:
                    float(count_str)
                except ValueError:
                    raise ValueError(
                        f"Row {i}, column {fields[0]}: "
                        f"expected numeric count, got '{count_str}'"
                    )
        except Exception as e:
            raise ValueError(f"Error validating expression file row {i}: {e}")
    
    return header[1:]  # Return sample names


def validate_metadata_file(filepath, sample_names, sample_suffix=None):
    """
    Validate metadata file and check compatibility with expression file.
    
    Parameters
    ----------
    filepath : str
        Path to metadata file.
    sample_names : list[str]
        Sample names from expression file.
    sample_suffix : str, optional
        Regex pattern to strip from sample names.
    
    Raises
    ------
    ValueError
        If metadata file is invalid or incompatible with expression file.
    """
    try:
        with open(filepath, 'r') as f:
            lines = f.readlines()
    except IOError as e:
        raise ValueError(f"Cannot read metadata file: {e}")
    
    if len(lines) < 2:
        raise ValueError("Metadata file must have at least 2 rows (header + 1 sample)")
    
    # Parse header (case-insensitive)
    header = lines[0].strip().split('\t')
    header_lower = [h.lower() for h in header]
    
    if 'sample_name' not in header_lower:
        raise ValueError(
            "Metadata file must have 'sample_name' column. "
            f"Found columns: {', '.join(header)}"
        )
    
    # Check for required group columns
    has_group = 'group' in header_lower
    has_species_tissue = ('species' in header_lower and 'tissue' in header_lower)
    
    if not (has_group or has_species_tissue):
        raise ValueError(
            "Metadata file must have either 'group' OR both 'species' and 'tissue' columns. "
            f"Found columns: {', '.join(header)}"
        )
    
    # Get sample_name column index
    sample_col = header_lower.index('sample_name')
    
    # Parse metadata samples and match to expression file
    metadata_samples = set()
    for i, line in enumerate(lines[1:], start=2):
        fields = line.strip().split('\t')
        if len(fields) <= sample_col:
            raise ValueError(f"Metadata row {i} has too few columns")
        metadata_samples.add(fields[sample_col])
    
    # Check sample matching (accounting for suffix stripping)
    stripped_sample_names = set(sample_names)
    if sample_suffix:
        stripped_sample_names = set(re.sub(sample_suffix, "", s) for s in sample_names)
    
    matching = metadata_samples & stripped_sample_names
    only_in_metadata = metadata_samples - stripped_sample_names
    only_in_expression = stripped_sample_names - metadata_samples
    
    if len(matching) == 0:
        raise ValueError(
            f"No samples match between expression file and metadata. "
            f"Expression samples: {', '.join(sorted(list(stripped_sample_names)[:5]))}... "
            f"Metadata samples: {', '.join(sorted(list(metadata_samples)[:5]))}..."
        )
    
    pct_matched = 100 * len(matching) / len(stripped_sample_names)
    if pct_matched < 90:
        print(
            f"WARNING: Only {pct_matched:.1f}% of expression samples match metadata. "
            f"In metadata but not expression: {', '.join(sorted(list(only_in_metadata)[:3]))}... "
            f"In expression but not metadata: {', '.join(sorted(list(only_in_expression)[:3]))}...",
            file=sys.stderr
        )


def build_metadata_from_groups(group_specs):
    """
    Parse --group-samples strings into a list of (sample_name, group) rows.

    Parameters
    ----------
    group_specs : list[str]
        Each element has the form "GroupLabel:s1,s2,s3".

    Returns
    -------
    list[tuple[str, str]]
        Rows of (sample_name, group) for every sample across all groups.
    """
    rows = []
    for spec in group_specs:
        if ":" not in spec:
            raise ValueError(
                f"--group-samples entry '{spec}' must be in the form "
                "'GroupLabel:sample1,sample2,...'"
            )
        label, samples_str = spec.split(":", 1)
        for s in samples_str.split(","):
            s = s.strip()
            if s:
                rows.append((s, label.strip()))
    return rows


def write_temp_metadata(rows, tmp_dir):
    """Write (sample_name, group) rows to a temp TSV and return its path."""
    fh = tempfile.NamedTemporaryFile(
        mode="w", suffix=".tsv", dir=tmp_dir, delete=False
    )
    writer = csv.writer(fh, delimiter="\t")
    writer.writerow(["sample_name", "group"])
    writer.writerows(rows)
    fh.close()
    return fh.name


def main(args):
    r_script = os.path.join(
        os.path.dirname(os.path.dirname(os.path.abspath(__file__))),
        "src", "R_src", "edger_de.R"
    )

    if not os.path.isfile(r_script):
        print(f"ERROR: R script not found at {r_script}", file=sys.stderr)
        sys.exit(1)

    # Validate Rscript executable ----------------------------------------
    if not os.path.isfile(args.rscript_executable):
        print(
            f"ERROR: Rscript executable not found at: {args.rscript_executable}\n"
            f"       Specify a different path with --rscript-executable",
            file=sys.stderr
        )
        sys.exit(1)

    # Validate parameter ranges ------------------------------------------
    if args.fdr_threshold < 0 or args.fdr_threshold > 1:
        print(
            f"ERROR: FDR threshold must be between 0 and 1 (got {args.fdr_threshold})",
            file=sys.stderr
        )
        sys.exit(1)
    
    if args.logfc_threshold < 0:
        print(
            f"ERROR: logFC threshold must be non-negative (got {args.logfc_threshold})",
            file=sys.stderr
        )
        sys.exit(1)

    # Validate expression file -------------------------------------------
    print("Validating expression file...", file=sys.stderr)
    try:
        sample_names = validate_expression_file(args.expression_file)
    except ValueError as e:
        print(f"ERROR: {e}", file=sys.stderr)
        sys.exit(1)

    # Resolve metadata path -----------------------------------------------
    temp_meta_path = None

    if args.metadata_file:
        if not os.path.isfile(args.metadata_file):
            print(f"ERROR: Metadata file not found: {args.metadata_file}",
                  file=sys.stderr)
            sys.exit(1)
        metadata_path = args.metadata_file
    elif args.group_samples:
        try:
            rows = build_metadata_from_groups(args.group_samples)
        except ValueError as e:
            print(f"ERROR: {e}", file=sys.stderr)
            sys.exit(1)
        os.makedirs(args.output_dir, exist_ok=True)
        temp_meta_path = write_temp_metadata(rows, args.output_dir)
        metadata_path = temp_meta_path
        print(f"Built temporary metadata file from --group-samples: {temp_meta_path}")
        # For inline groups, skip metadata validation (metadata is auto-generated and correct)
        sample_names_to_check = set()
        for row in rows:
            sample_names_to_check.add(row[0])
        if not sample_names_to_check <= set(sample_names):
            missing = sample_names_to_check - set(sample_names)
            print(
                f"WARNING: Samples in --group-samples not in expression file: "
                f"{', '.join(sorted(missing))}",
                file=sys.stderr
            )
    else:
        print(
            "ERROR: Provide either --metadata-file or --group-samples.",
            file=sys.stderr
        )
        sys.exit(1)

    # Validate metadata file (if provided, not auto-generated) -----------
    if args.metadata_file:
        print("Validating metadata file...", file=sys.stderr)
        try:
            validate_metadata_file(
                metadata_path, 
                sample_names, 
                sample_suffix=args.sample_suffix
            )
        except ValueError as e:
            print(f"ERROR: {e}", file=sys.stderr)
            sys.exit(1)

    os.makedirs(args.output_dir, exist_ok=True)

    # Build R command ------------------------------------------------------
    sample_suffix = args.sample_suffix if args.sample_suffix else "NULL"

    cmd = [
        args.rscript_executable,
        r_script,
        args.expression_file,
        args.output_dir,
        metadata_path,
        str(args.fdr_threshold),
        str(args.logfc_threshold),
        sample_suffix,
    ]

    print("=" * 70)
    print("SalmonStreamer EdgeRDE")
    print("=" * 70)
    print(f"Expression file : {args.expression_file}")
    print(f"Metadata file   : {metadata_path}")
    print(f"Output dir      : {args.output_dir}")
    print(f"FDR threshold   : {args.fdr_threshold}")
    print(f"logFC threshold : {args.logfc_threshold}")
    if args.sample_suffix:
        print(f"Sample suffix   : {args.sample_suffix}")
    print()

    try:
        result = subprocess.run(cmd, check=True, capture_output=False)
    except subprocess.CalledProcessError as e:
        print(
            f"ERROR: R script failed with exit code {e.returncode}\n"
            f"       Check the error messages above for details.",
            file=sys.stderr
        )
        sys.exit(e.returncode)
    except FileNotFoundError as e:
        print(
            f"ERROR: Cannot find Rscript executable: {args.rscript_executable}\n"
            f"       {e}",
            file=sys.stderr
        )
        sys.exit(1)
    finally:
        if temp_meta_path and os.path.isfile(temp_meta_path):
            os.remove(temp_meta_path)

    print("\nEdgeRDE complete.")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description=(
            "Run edgeR-based differential expression analysis "
            "(SalmonStreamer EdgeRDE subcommand)."
        ),
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Metadata file required columns:
  sample_name   Matches count matrix column headers (after --sample-suffix).
  group         Experimental group label (or: species + tissue columns).

Examples:
  # Using a metadata file
  SalmonStreamer.py EdgeRDE \\
    --expression-file counts.tsv \\
    --output-dir DE_results/ \\
    --metadata-file metadata.tsv

  # Using inline group specification
  SalmonStreamer.py EdgeRDE \\
    --expression-file counts.tsv \\
    --output-dir DE_results/ \\
    --group-samples SpeciesA:s1,s2,s3 SpeciesB:s4,s5,s6

  # Strip a common suffix from sample names before matching metadata
  SalmonStreamer.py EdgeRDE \\
    --expression-file counts.tsv \\
    --metadata-file metadata.tsv \\
    --output-dir DE_results/ \\
    --sample-suffix "_R1_filtered"
        """
    )

    parser.add_argument(
        "--expression-file", required=True,
        help="Tab-separated count matrix (genes x samples, first column = gene IDs)."
    )
    parser.add_argument(
        "--output-dir", required=True,
        help="Directory for all output files."
    )
    parser.add_argument(
        "--metadata-file", default=None,
        help=(
            "Tab-separated sample metadata. Required columns: sample_name, group "
            "(or species + tissue). Takes priority over --group-samples."
        )
    )
    parser.add_argument(
        "--group-samples", nargs="+", default=None, metavar="GROUP:s1,s2,...",
        help=(
            "Inline group specification when no metadata file is available. "
            "Format: 'GroupLabel:sample1,sample2,...' (one entry per group). "
            "Example: --group-samples SpeciesA:s1,s2 SpeciesB:s3,s4"
        )
    )
    parser.add_argument(
        "--fdr-threshold", type=float, default=0.05,
        help="FDR significance cutoff (default: 0.05)."
    )
    parser.add_argument(
        "--logfc-threshold", type=float, default=1.0,
        help="|log2FC| threshold for significance categories (default: 1.0)."
    )
    parser.add_argument(
        "--sample-suffix", default=None,
        help=(
            "Regex pattern stripped from count matrix column names before "
            "matching to metadata sample_name values "
            "(e.g. '_R1_filtered$'). Optional."
        )
    )
    parser.add_argument(
        "--rscript-executable",
        default=os.path.expanduser("~/.conda/envs/PyR/bin/Rscript"),
        help="Path to the Rscript executable (default: ~/.conda/envs/PyR/bin/Rscript)."
    )

    main(parser.parse_args())

File: src/R_src/edger_de.R
#!/usr/bin/env Rscript

################################################################################
# Differential Expression Analysis using edgeR
# SalmonStreamer Pipeline Module
#
# Author: Luis Javier Madrigal-Roca
# Date: 2026-04-30
#
# Usage (called by SalmonStreamer EdgeRDE):
#   Rscript edger_de.R <input_file> <output_dir> <metadata_file> \
#                      [fdr_threshold] [logfc_threshold] [sample_suffix]
#
# Arguments:
#   input_file       Tab-separated count matrix (genes x samples).
#                    First column: gene/transcript IDs.
#                    Remaining columns: one per sample (column header = sample name).
#
#   output_dir       Directory where all output files are written.
#
#   metadata_file    Tab-separated file describing samples.
#
#                    REQUIRED COLUMNS:
#                      sample_name  Identifier matching the count matrix column
#                                   header (after sample_suffix stripping, if any).
#                      group        Experimental group label used to define
#                                   contrasts (e.g. "SpeciesA_Leaf").
#
#                    ALTERNATIVE: provide 'species' + 'tissue' instead of
#                    'group'; the script will derive group as
#                    paste(species, tissue, sep = "_").
#
#                    OPTIONAL COLUMNS (ignored by this script, may be present):
#                      platform, library, flowcell, lane, replicate, ...
#
#   fdr_threshold    FDR significance cutoff (default: 0.05).
#   logfc_threshold  |log2FC| threshold for significance categories (default: 1).
#   sample_suffix    Regex stripped from count matrix column names before
#                    matching to metadata sample_name values.
#                    Example: "_R1_filtered$". Pass "NULL" to skip.
################################################################################

options(bitmapType = "cairo")
if (!interactive()) options(device = "pdf")

suppressPackageStartupMessages({
  library(edgeR)
  library(limma)
  library(ggplot2)
  library(pheatmap)
  library(RColorBrewer)
  library(dplyr)
  library(readr)
})

# ---------------------------------------------------------------------------
# 0. Argument parsing
# ---------------------------------------------------------------------------

args <- commandArgs(trailingOnly = TRUE)
if (length(args) < 3) {
  stop(
    "Usage: edger_de.R <input_file> <output_dir> <metadata_file> ",
    "[fdr_threshold] [logfc_threshold] [sample_suffix]"
  )
}

input_file      <- args[1]
output_dir      <- args[2]
metadata_arg    <- args[3]
fdr_threshold   <- if (length(args) >= 4) as.numeric(args[4]) else 0.05
logfc_threshold <- if (length(args) >= 5) as.numeric(args[5]) else 1.0
sample_suffix   <- if (length(args) >= 6 && args[6] != "NULL") args[6] else NULL

metadata_file <- if (metadata_arg == "NULL") NULL else metadata_arg

dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)

cat("=== SalmonStreamer EdgeRDE Analysis ===\n")
cat("Input file     :", input_file, "\n")
cat("Output dir     :", output_dir, "\n")
cat("Metadata file  :", ifelse(is.null(metadata_file), "(none)", metadata_file), "\n")
cat("FDR threshold  :", fdr_threshold, "\n")
cat("logFC threshold:", logfc_threshold, "\n\n")

# ---------------------------------------------------------------------------
# 1. Load count matrix
# ---------------------------------------------------------------------------

if (!file.exists(input_file)) stop("Input file not found: ", input_file)

counts_data    <- read_tsv(input_file, show_col_types = FALSE)
transcript_ids <- counts_data[[1]]
count_matrix   <- as.matrix(counts_data[, -1])
rownames(count_matrix) <- transcript_ids

cat("Data dimensions:", nrow(count_matrix), "genes x", ncol(count_matrix), "samples\n")

if (!is.null(sample_suffix)) {
  colnames(count_matrix) <- gsub(sample_suffix, "", colnames(count_matrix))
  cat("Stripped suffix '", sample_suffix, "' from sample names\n", sep = "")
}
sample_names <- colnames(count_matrix)

# ---------------------------------------------------------------------------
# 2. Load and validate metadata
# ---------------------------------------------------------------------------

if (is.null(metadata_file)) {
  stop(
    "A metadata file is required. Provide it via --metadata-file or ",
    "use --group-samples to let the Python wrapper build one automatically."
  )
}
if (!file.exists(metadata_file)) stop("Metadata file not found: ", metadata_file)

metadata <- read_tsv(metadata_file, show_col_types = FALSE)

# Normalise the sample identifier column name
col_lower <- tolower(colnames(metadata))
if ("sample_name" %in% col_lower) {
  colnames(metadata)[col_lower == "sample_name"] <- "Sample"
} else if (!"sample" %in% col_lower) {
  stop("Metadata must contain a 'sample_name' or 'Sample' column.")
} else {
  colnames(metadata)[col_lower == "sample"] <- "Sample"
}

# Normalise / derive the group column
col_lower <- tolower(colnames(metadata))
if ("group" %in% col_lower) {
  colnames(metadata)[col_lower == "group"] <- "Group"
} else if (all(c("species", "tissue") %in% col_lower)) {
  colnames(metadata)[col_lower == "species"] <- "Species"
  colnames(metadata)[col_lower == "tissue"]  <- "Tissue"
  metadata$Group <- paste(metadata$Species, metadata$Tissue, sep = "_")
  cat("Derived Group from species + tissue columns\n")
} else {
  stop(
    "Metadata must contain either a 'group' column or both ",
    "'species' and 'tissue' columns."
  )
}

# Align metadata to count matrix
in_both   <- intersect(metadata$Sample, sample_names)
only_meta <- setdiff(metadata$Sample, sample_names)
only_mat  <- setdiff(sample_names, metadata$Sample)

if (length(only_meta) > 0)
  warning("Samples in metadata but not in count matrix: ",
          paste(only_meta, collapse = ", "))
if (length(only_mat) > 0)
  warning("Samples in count matrix but not in metadata: ",
          paste(only_mat, collapse = ", "))
if (length(in_both) == 0)
  stop("No samples shared between metadata and count matrix. ",
       "Check sample_name values and any --sample-suffix stripping.")

metadata     <- metadata[metadata$Sample %in% in_both, ]
count_matrix <- count_matrix[, metadata$Sample, drop = FALSE]

cat("Samples retained:", nrow(metadata), "\n")
cat("Groups          :", paste(unique(metadata$Group), collapse = ", "), "\n\n")

# ---------------------------------------------------------------------------
# 3. DGEList, filtering, TMM normalisation
# ---------------------------------------------------------------------------

metadata$group_factor <- factor(metadata$Group)
dge <- DGEList(counts = count_matrix, samples = metadata,
               group = metadata$group_factor)

cat("Genes before filtering:", nrow(dge), "\n")
keep <- filterByExpr(dge, group = metadata$group_factor)
dge  <- dge[keep, , keep.lib.sizes = FALSE]
cat("Genes after filterByExpr:", nrow(dge), "\n")

dge <- calcNormFactors(dge, method = "TMM")
cat("TMM factors:", paste(round(dge$samples$norm.factors, 3), collapse = ", "), "\n\n")

# ---------------------------------------------------------------------------
# 4. Exploratory plots
# ---------------------------------------------------------------------------

cat("Generating exploratory plots...\n")
lcpm <- cpm(dge, log = TRUE)

# PCA
pca_result <- prcomp(t(lcpm), scale. = FALSE)
pct_var    <- round(summary(pca_result)$importance[2, 1:2] * 100, 1)

pca_df <- data.frame(
  PC1    = pca_result$x[, 1],
  PC2    = pca_result$x[, 2],
  Sample = rownames(pca_result$x)
)
pca_df <- merge(pca_df, metadata[, c("Sample", "Group")], by = "Sample")
hulls  <- pca_df %>% group_by(Group) %>% slice(chull(PC1, PC2))

p_pca <- ggplot(pca_df, aes(x = PC1, y = PC2, color = Group, fill = Group)) +
  geom_polygon(data = hulls, alpha = 0.15, show.legend = FALSE) +
  geom_point(size = 3, alpha = 0.85) +
  labs(
    title = "PCA of Gene Expression",
    x     = paste0("PC1 (", pct_var[1], "%)"),
    y     = paste0("PC2 (", pct_var[2], "%)")
  ) +
  theme_bw() +
  theme(legend.title = element_blank(), panel.grid = element_blank())

ggsave(file.path(output_dir, "PCA_plot.pdf"), p_pca,
       width = 8, height = 6, device = "pdf")
ggsave(file.path(output_dir, "PCA_plot.png"), p_pca,
       width = 8, height = 6, dpi = 300, device = "png")

# Sample correlation heatmap
cor_matrix <- cor(lcpm, method = "pearson")
ann_df     <- data.frame(Group = metadata$Group, row.names = metadata$Sample)

pdf(file.path(output_dir, "sample_correlation_heatmap.pdf"), width = 10, height = 10)
pheatmap(cor_matrix, annotation_col = ann_df, annotation_row = ann_df,
         main = "Sample Correlation Heatmap")
dev.off()

# Library sizes
lib_df <- data.frame(
  Sample  = metadata$Sample,
  LibSize = dge$samples$lib.size / 1e6,
  Group   = metadata$Group
)
p_lib <- ggplot(lib_df, aes(x = Sample, y = LibSize, fill = Group)) +
  geom_col(alpha = 0.8) +
  labs(title = "Library Sizes", x = "Sample", y = "Library Size (Millions)") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        panel.grid   = element_blank())

ggsave(file.path(output_dir, "library_sizes.pdf"), p_lib,
       width = 10, height = 6, device = "pdf")
ggsave(file.path(output_dir, "library_sizes.png"), p_lib,
       width = 10, height = 6, dpi = 300, device = "png")

# ---------------------------------------------------------------------------
# 5. Statistical analysis — edgeR quasi-likelihood pipeline
# ---------------------------------------------------------------------------

cat("Fitting edgeR quasi-likelihood model...\n")

# ~ 0 + group_factor gives one explicit column per group (no reference level),
# which makes all pairwise contrasts symmetric and equally easy to specify.
design <- model.matrix(~ 0 + group_factor, data = metadata)
colnames(design) <- make.names(levels(metadata$group_factor))

dge <- estimateDisp(dge, design, robust = TRUE)
cat("Common BCV:", round(sqrt(dge$common.dispersion), 4), "\n")

pdf(file.path(output_dir, "dispersion_plot.pdf"), width = 8, height = 6)
plotBCV(dge, main = "Biological Coefficient of Variation")
dev.off()

fit <- glmQLFit(dge, design, robust = TRUE)

pdf(file.path(output_dir, "QL_dispersion_plot.pdf"), width = 8, height = 6)
plotQLDisp(fit, main = "Quasi-likelihood Dispersion")
dev.off()

# All pairwise contrasts
groups      <- colnames(design)
orig_groups <- levels(metadata$group_factor)  # keep original labels for naming
cat("Groups:", paste(orig_groups, collapse = ", "), "\n\n")

results_list <- list()
for (i in seq_len(length(groups) - 1)) {
  for (j in (i + 1):length(groups)) {
    g1     <- groups[i]
    g2     <- groups[j]
    g1_lab <- orig_groups[i]
    g2_lab <- orig_groups[j]
    cmp    <- paste0(g1_lab, "_vs_", g2_lab)

    contrast_vec       <- setNames(numeric(length(groups)), groups)
    contrast_vec[g1]   <- 1
    contrast_vec[g2]   <- -1

    qlf <- glmQLFTest(fit, contrast = contrast_vec)
    results_list[[cmp]] <- topTags(qlf, n = Inf)$table
    cat("Completed:", cmp, "\n")
  }
}

# ---------------------------------------------------------------------------
# 6. Results export and visualisations
# ---------------------------------------------------------------------------

cat("\nExporting results...\n")
all_sig_genes <- character(0)

for (cmp in names(results_list)) {
  tbl <- results_list[[cmp]]
  tbl$TranscriptID  <- rownames(tbl)
  tbl <- tbl[, c("TranscriptID", setdiff(colnames(tbl), "TranscriptID"))]

  tbl$Significance <- "Not Significant"
  tbl$Significance[tbl$FDR < fdr_threshold &
                     abs(tbl$logFC) > logfc_threshold] <- "Significant"
  tbl$Significance[tbl$FDR < 0.01 &
                     abs(tbl$logFC) > 2 * logfc_threshold] <- "Highly Significant"
  tbl <- tbl[order(tbl$FDR), ]

  write_tsv(tbl, file.path(output_dir, paste0(cmp, "_DE_results.tsv")))

  sig <- tbl[tbl$FDR < fdr_threshold, ]
  write_tsv(sig, file.path(output_dir, paste0(cmp, "_significant_genes.tsv")))
  all_sig_genes <- union(all_sig_genes, sig$TranscriptID)

  cat(sprintf("  %s -> tested: %d | sig: %d | up: %d | down: %d\n",
              cmp, nrow(tbl), nrow(sig),
              sum(sig$logFC > 0), sum(sig$logFC < 0)))

  # Volcano plot
  vdata <- tbl
  vdata$neg_log10_FDR <- -log10(pmax(vdata$FDR, .Machine$double.xmin))

  p_vol <- ggplot(vdata, aes(x = logFC, y = neg_log10_FDR, color = Significance)) +
    geom_point(alpha = 0.5, size = 0.8) +
    scale_color_manual(values = c(
      "Not Significant"    = "grey70",
      "Significant"        = "#377EB8",
      "Highly Significant" = "#E41A1C"
    )) +
    geom_vline(xintercept = c(-logfc_threshold, logfc_threshold),
               linetype = "dashed", alpha = 0.5) +
    geom_hline(yintercept = -log10(fdr_threshold),
               linetype = "dashed", alpha = 0.5) +
    labs(
      title = paste("Volcano:", cmp),
      x     = "log2 Fold Change",
      y     = paste0("-log10(FDR)")
    ) +
    theme_bw() +
    theme(legend.title = element_blank(), panel.grid = element_blank())

  ggsave(file.path(output_dir, paste0(cmp, "_volcano.pdf")), p_vol,
         width = 10, height = 8, device = "pdf")
  ggsave(file.path(output_dir, paste0(cmp, "_volcano.png")), p_vol,
         width = 10, height = 8, dpi = 300, device = "png")
}

# Expression heatmap of top DE genes
if (length(all_sig_genes) > 0) {
  genes_to_plot <- if (length(all_sig_genes) > 100) {
    unique(unlist(lapply(results_list, function(x)
      head(rownames(x)[x$FDR < fdr_threshold], 50))))
  } else {
    all_sig_genes
  }
  genes_to_plot <- intersect(genes_to_plot, rownames(lcpm))

  if (length(genes_to_plot) > 0) {
    hz <- t(scale(t(lcpm[genes_to_plot, , drop = FALSE])))
    pdf(file.path(output_dir, "DE_genes_heatmap.pdf"),
        width = 12, height = max(8, length(genes_to_plot) * 0.15))
    pheatmap(hz, annotation_col = ann_df,
             show_rownames = length(genes_to_plot) <= 50,
             main = "Top Differentially Expressed Genes",
             fontsize_row = 6)
    dev.off()
  }
}

# ---------------------------------------------------------------------------
# 7. Summary report
# ---------------------------------------------------------------------------

summary_lines <- c(
  "=== SalmonStreamer EdgeRDE Analysis Summary ===",
  paste("Date         :", Sys.Date()),
  paste("Input file   :", input_file),
  paste("Metadata     :", metadata_file),
  paste("Genes (post-filter):", nrow(dge)),
  paste("Samples      :", ncol(dge)),
  paste("Groups       :", paste(orig_groups, collapse = ", ")),
  paste("FDR threshold:", fdr_threshold),
  paste("logFC threshold:", logfc_threshold),
  "",
  "Comparisons:"
)
for (cmp in names(results_list)) {
  tbl <- results_list[[cmp]]
  sig <- sum(tbl$FDR < fdr_threshold, na.rm = TRUE)
  summary_lines <- c(summary_lines,
    sprintf("  %-60s %d significant genes", cmp, sig))
}

writeLines(summary_lines, file.path(output_dir, "analysis_summary.txt"))
writeLines(capture.output(sessionInfo()), file.path(output_dir, "session_info.txt"))

cat("\nDone. Results written to:", output_dir, "\n")
